### ISSUE TITLE
Authorized Cluster Endpoint (ACE) support for imported RKE2/K3S clusters

### ISSUE THREAD

**Issue by thedadams** [rancher/rancher#35161]
When provisioning an RKE2/K3S cluster from Rancher, it is possible to use Authorized Cluster Endpoint on the downstream cluster. It would be good to have the same support for imported RKE2/K3S clusters.

A few notes about the support:
1. This must work for RKE2/K3S clusters imported in 2.6.x and those imported in a previous version of Rancher and then Rancher upgraded to 2.6.x
2. Manual steps to alter the config of the downstream RKE2/K3S cluster and deploy the `kube-api-authn-webhook` are expected. These steps should be documented.
3. If a future implementation of ACE requires an update to the `kube-api-authn-webhook`, then this would also have to be done manually.

JIRA ID: SURE-3474

**Comment by thedadams** [comment]
UI issue: https://github.com/rancher/dashboard/issues/4410

**Comment by davidnuzik** [comment]
Note: I filed https://github.com/rancher/rancher/issues/35338 while testing this issue.

**Comment by davidnuzik** [comment]
**My checks PASSED**
Please note, as Donnie mentioned in a prior comment a few manual steps are required. This will get documented via the docs issue he created, also mentioned above (https://github.com/rancher/docs/issues/3612). Further, Donnie mentioned above that [the UI is not yet completed](https://github.com/rancher/dashboard/issues/4410) (as of the time of me writing this), and therefore testing was done via the kuberenetes API on the `local` cluster by modifying the cluster (clusters.provisioning.cattle.io) for each imported cluster to enable localClusterAuthEndpoint

When the UI issue is completed later and ready for test, we will test this again but via the UI. This will also enable us to test supplying an FQDN and Cert. It was not trivial to test the FQDN and Cert fields prior to the UI change, so it was agreed it would make sense to test this only after the UI change is in before release of v2.6.3.

Note, an internal test plan was also created for future reference and was adequately reviewed.
 
**Reproduction Steps:**

Not required. New feature/enhancement.

-----
 
**Validation Environment:**

Rancher version: v2.6-head `cc2b8be` 11/11/21
Rancher cluster type: HA Install
 
Downstream cluster type: imported k3s and imported rke2
Downstream K8s version:  v1.20.12+rke2r1 and  v1.20.12+k3s1 (post rancher upgrade, enable and ensure ace works, upgrade each to  v1.21.6+rke2r1 and  v1.21.6+k3s1 and ensure ACE still works)
 
**Validation steps:**

1. Install HA Rancher 2.5.11. (Latest 2.5.x available to me at the time)
2. Create a standalone K3S cluster with version v1.20.12+k3s1. Import it into Rancher.
Create a standalone RKE2 cluster with version v1.20.12+rke2r1. Import it into Rancher.
3. Download the kubeconfigs for each imported cluster in Rancher. There is no 2nd context, as expected (ACE is not enabled). Confirmed can access the clusters with the primary (Rancher) context.
4. Upgrade HA Rancher to v2.6-head (`cc2b8be` 11/11/2021).
5. Enable ACE on each imported cluster. These steps are unique to this issue and won't be required later once the frontend is implemented. These steps are outlined below in the "Additional Info" section.
6. On each K3S and RKE2 cluster (any/all controlplane nodes), add the webhook following Donnie's steps [here](https://github.com/rancher/rancher/issues/35161#issuecomment-951335144) (near the bottom of this comment). This is a mandatory manual step and will get documented as mentioned above.
7. Now that ACE is enabled on the imported K3S and RKE2 clusters and the webhooks are setup, download the kubeconfigs for each. Leverage them remotely with a kubectl client. Switch to the 2nd context (`kubectl config use-context xyz`). Confirm you get a response to kubectl queries. Take care not to execute any prior queries as kubectl client can cache queries. This works.
8. Now, to be absolutely certain this is working correctly and that we are indeed using the secondary (ACE) context, shutdown the rancher server - scale the rancher pods to zero or shutdown the nodes temporarily. Confirm you can still access each imported cluster with Rancher down when using the 2nd (ACE) context. This works.
9. Last, in the Rancher UI, upgrade each imported cluster (edit the cluster in the UI) to the latest available at the time of me writing this this is version v1.21.6. Let the upgrade complete and ensure the secondary (ACE) context still works. This works.
 
**Additional Info:**
How to enable localClusterAuthEndpoint before the UI is in:
First: Do this post-upgrade, not before.
1. On the local rancher cluster, get kubectl shell access. Execute the following to edit each imported cluster object:
`kubectl -n fleet-default edit clusters.provisioning.cattle.io NAME-OF-YOUR-IMPORTED-K3S-OR-RKE2-CLUSTER`
This is the cluster name as the ember UI used to name them so it will be something like c-randstring (e.g. something like c-4b10ca or similar)
2. in the "spec:" section add an entry like:
```
spec:
  localClusterAuthEndpoint:
    enabled: true
```
3. write and quite out (vim command) and this will update the object. ACE is now enabled on this imported cluster you specified.
4. Don't forget to do it for each cluster. Don't forget that you need to set up the webhook as well. And -- to repeat, you shouldn't have to do this once the UI change is in.

**Comment by thedadams** [comment]
The case of an imported cluster in Rancher v2.6+ needs to be retested with the ACE config added to the management cluster instead of the provisioning cluster.

**Comment by davidnuzik** [comment]
v2.5.11 HA Rancher install upgraded to v2.6-head `da2e59c` 12/7/2021

**My checks PASSED**

I executed the exact same steps as last time (see my [comment](https://github.com/rancher/rancher/issues/35161#issuecomment-967262585) here)

EXCEPT:

I enabled ACE on the management cluster this time, rather than on the provisioning cluster object (see additional info section near end of that referenced comment).
This can be accomplished by either modifying the clusters.management.cattle.io (cluster) object via kubectl on the local rancher cluster or directly in the UI now by simply editing the cluster and enabling ACE that way. I tried both and both work without any issues. I was successfully able to use the secondary context without any issues.
